runtime/gc_vlhgc/RememberedSetCardBucket.cpp

```diff
@@ -213,13 +214,14 @@ MM_RememberedSetCardBucket::compact(MM_EnvironmentVLHGC *env)
 
 			/* find top index for this buffer */
 			UDATA fromIndexTop = MAX_BUFFER_SIZE;
-			if (isCurrentSlotWithinBuffer(fromBufferCardList)) {
-				fromIndexTop = _current -  fromBufferCardList;
+			if (isCurrentSlotWithinBuffer(env, fromBufferCardList)) {
+				fromIndexTop = env->subtractCardAddresses(_current, fromBufferCardList);
 			}
 
 			for (UDATA fromIndex = 0; fromIndex < fromIndexTop; fromIndex++) {
-				if (0 != fromBufferCardList[fromIndex]) {
-					toBufferCardList[toIndex++] = fromBufferCardList[fromIndex];
+				MM_RememberedSetCard card = env->readCard(env->addToCardAddress(fromBufferCardList, fromIndex));
+				if (0 != card) {
```

Comment:
Sooner or later tight loops like this should be optimized so that 'is CR' check done ideally only once per iteration. I can see 4 hidden checks here.